### ISSUE TITLE
Private note - Hmm, It's Not Here' Displayed When Opening Private Note Offline After Previously Access online

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#39507]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.59-3
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:** N/A
**Email or phone of affected tester (no customers):** abebemiherat@gmail.com
**Issue reported by:** Applause - Internal Team

## Action Performed:
1, Navigate to any chat and click on the header
2, Open a private note
3, Make it offline
4, Click on the header again and try to open the private note
5, Observe the display of the 'Hmm, It's Not Here' page

## Expected Result:
The loading indicator should appear instead of displaying the 'Hmm, It's Not Here' page.
Alternatively, the private note should be accessible offline

## Actual Result:
When attempting to access the private note offline after previously accessing it online, the 'Hmm, It's Not Here' page is displayed

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

</details> 

https://github.com/Expensify/App/assets/78819774/59d229da-0687-4c47-b798-ee7ff33e7a08

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by lanitochka17** [comment]
@CortneyOfstad  FYI I haven't added the External label as I wasn't 100% sure about this issue. Please take a look and add the label if you agree it's a bug and can be handled by external contributors

**Comment by shahinyan11** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Private note - Hmm, It's Not Here' Displayed When Opening Private Note Offline After Previously Access online
### What is the root cause of that problem?
Du to [this](https://github.com/Expensify/App/blob/7177173f4b686a32bfcb94f07f0541b2f865f31b/src/pages/home/report/withReportAndPrivateNotesOrNotFound.tsx#L58C1-L72C1) implementation the `shouldShowNotFoundPage` will be always `true` if the user is offline
### What changes do you think we should make in order to solve the problem?
Change [this](https://github.com/Expensify/App/blob/7177173f4b686a32bfcb94f07f0541b2f865f31b/src/pages/home/report/withReportAndPrivateNotesOrNotFound.tsx#L55) line like below
```
const shouldShowFullScreenLoadingIndicator = !isPrivateNotesFetchFinished || isOffline;
```

### What alternative solutions did you explore? (Optional)

**Comment by FitseTLT** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
 Hmm, It's Not Here' Displayed When Opening Private Note Offline After Previously Access online

### What is the root cause of that problem?
When notes are empty and not loading we are showing not found page because of this line
https://github.com/Expensify/App/blob/398636a109452a6442d5e9fd0774f928ceaddee0/src/pages/home/report/withReportAndPrivateNotesOrNotFound.tsx#L69-L71
### What changes do you think we should make in order to solve the problem?

We should change it to 
```
// As notes being empty and not loading is a valid case, show not found view only in offline mode.
                return false;
            }, [report, isOtherUserNote, shouldShowFullScreenLoadingIndicator, isPrivateNotesEmpty, isReconnecting]);
```
### What alternative solutions did you explore? (Optional)

**Comment by shahinyan11** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/39507#issuecomment-2034919348). A new solution has been added as the main one and the previous one has been moved to the alternative

**Comment by CortneyOfstad** [comment]
@c3024 we have some proposals above for your review — thank you!

**Comment by ahmedGaber93** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Private note - Hmm, It's Not Here' Displayed When Opening Private Note Offline After Previously Access online

### What is the root cause of that problem?
The code designed to return early `LoadingPage` then `NotFoundPage`
https://github.com/Expensify/App/blob/7177173f4b686a32bfcb94f07f0541b2f865f31b/src/pages/home/report/withReportAndPrivateNotesOrNotFound.tsx#L73-L79
the error in `shouldShowFullScreenLoadingIndicator` calculation, exactly in `isPrivateNotesFetchTriggered`.
https://github.com/Expensify/App/blob/7177173f4b686a32bfcb94f07f0541b2f865f31b/src/pages/home/report/withReportAndPrivateNotesOrNotFound.tsx#L38

This value defined by condition  `report.isLoadingPrivateNotes !== undefined`, but `report.isLoadingPrivateNotes` will be  undefined only one time for first time fetched forever in this device/browser, and after first API request it will set to true/false and will save in `Onyx` by this method [getReportPrivateNote](https://github.com/Expensify/App/blob/7177173f4b686a32bfcb94f07f0541b2f865f31b/src/libs/actions/Report.ts#L2682-L2710) 
### What changes do you think we should make in order to solve the problem?
we need to change  `isPrivateNotesFetchTriggered` definition to
```javascript
 const isPrivateNotesFetchTriggered = report.isLoadingPrivateNotes !== undefined; 
 const isPrivateNotesFetchTriggered = Boolean(report.isLoadingPrivateNotes);

```


### What alternative solutions did you explore? (Optional)
we can set `isLoadingPrivateNotes` to `undeifined` in [getReportPrivateNote](https://github.com/Expensify/App/blob/7177173f4b686a32bfcb94f07f0541b2f865f31b/src/libs/actions/Report.ts#L2682-L2710) in successData and failureData

**Comment by xzunknownzx** [comment]
Here are my _no-strings attached_ **free thoughts on the proposed solutions:**

@GandalfGwaihir's proposal to add an `isOffline` check to `isPrivateNotesFetchFinished` is a simple and effective way to address the root cause. This would prevent the "Not Here" page from incorrectly showing when offline.

@shahinyan11's alternative of updating `shouldShowFullScreenLoadingIndicator` similarly is also a clean solution. Either approach of incorporating the offline check would work.

@abzokhattab accurately identified that the empty string check in `isPrivateNotesEmpty` is faulty. Their fix to check for undefined instead is spot on. This handles the edge case of an empty note string.

@ahmedGaber93 makes a good observation that `isPrivateNotesFetchTriggered` will only be undefined once. Changing this to check the `boolean` value is cleaner.

--

Overall, I think @GandalfGwaihir's and @shahinyan11's solutions are the most straightforward fixes to address the core offline loading issue. @abzokhattab's fix for the empty note case is also important. @ahmedGaber93's `boolean` check refactor is a nice optimization.

In summary, there are several solid solutions proposed that identify the key issues and provide targeted improvements as requested in the offered listing for proposals... That said, I would focus on @GandalfGwaihir's offline check, @abzokhattab's empty note fix, and @ahmedGaber93's `boolean` refactor as the highest priorities. The contributors clearly understood the problem and have provided actionable solutions that will solve the issues at hand.

[Upwork Profile Here](https://shorturl.at/cilAD)

**Comment by xzunknownzx** [comment]
Looks like I was a little late to the party!
Well anyways, here are my _no-strings attached_ **free thoughts on the proposed solutions:**

@GandalfGwaihir's proposal to add an `isOffline` check to `isPrivateNotesFetchFinished` is a simple and effective way to address the root cause. This would prevent the "Not Here" page from incorrectly showing when offline.

@shahinyan11's alternative of updating `shouldShowFullScreenLoadingIndicator` similarly is also a clean solution. Either approach of incorporating the offline check would work.

@abzokhattab accurately identified that the empty string check in `isPrivateNotesEmpty` is faulty. Their fix to check for undefined instead is spot on. This handles the edge case of an empty note string.

@ahmedGaber93 makes a good observation that `isPrivateNotesFetchTriggered` will only be undefined once. Changing this to check the `boolean` value is cleaner.

--

Overall, I think @GandalfGwaihir's and @shahinyan11's solutions are the most straightforward fixes to address the core offline loading issue. @abzokhattab's fix for the empty note case is also important. @ahmedGaber93's `boolean` check refactor is a nice optimization.

In summary, there are several solid solutions proposed that identify the key issues and provide targeted improvements as requested in the offered listing for proposals... That said, I would focus on @GandalfGwaihir's offline check, @abzokhattab's empty note fix, and @ahmedGaber93's `boolean` refactor as the highest priorities. The contributors clearly understood the problem and have provided actionable solutions that will solve the issues at hand.

[Upwork Profile Here](https://shorturl.at/cilAD)

**Comment by c3024** [comment]
Thanks all for your proposals.

Once we open a private note online, that private note should be able to be seen without a loading indicator or a not found page when accessed later in offline.
This works fine if the note is a non-empty string. This behaviour should be the same for an empty string.

@GandalfGwaihir 's solution shows a loading indicator during offline even for a private note that was fetched earlier. IMO this is not desirable.


This comment here

https://github.com/Expensify/App/blob/ae6cc4927ce57e5be59e024109436909ed7e1bb5/src/pages/home/report/withReportAndPrivateNotesOrNotFound.tsx#L69

suggests we show not found page in case where we open a report details page offline before ever opening it online. This behaviour will be broken by the solution suggested by @shahinyan11 

RCA and solution suggested by @abzokhattab [here](https://github.com/Expensify/App/issues/39507#issuecomment-2035493277) looks like the correct way to fix this.

🎀 👀 🎀 C+ Reviewed

**Comment by CortneyOfstad** [comment]
@abzokhattab can you provide an update by EOD on where we're at with the PR? Thank you!

**Comment by CortneyOfstad** [comment]
Deployed to production 2 days ago! Adjusting the title since it did not update for some reason.
